<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jes" />


<title>Introduction to ggplot2 and dplyr</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>




<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}

.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="00-computer-setup.html">Computer Setup</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-A-R-intro.html">Intro to R</a>
    </li>
    <li>
      <a href="01-B-Rmarkdown-intro.html">R markdown</a>
    </li>
    <li>
      <a href="01-C-R-workshop.html">R workshop</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="02-A-tidyr.html">ggplot2 and tidyr</a>
    </li>
    <li>
      <a href="02-B-git-intro.html">Intro to git</a>
    </li>
    <li>
      <a href="02-C-student-projects.html">Project introductions</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-A-data-exploration.html">Data exploration</a>
    </li>
    <li>
      <a href="03-B-linear-models.html">Linear models</a>
    </li>
    <li>
      <a href="03-C-heterogeneity.html">Heterogeneity</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W4
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-A-mixed-models.html">Mixed effects models</a>
    </li>
    <li>
      <a href="04-B-binary-data.html">Binary &amp; proportional data</a>
    </li>
    <li>
      <a href="04-C-zero-data.html">Zero-inflated data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W5
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="05-A-Bayesian-linear-models.html">Bayesian linear models</a>
    </li>
    <li>
      <a href="05-B-Bayesian-priors.html">Bayesian inference with prior information</a>
    </li>
    <li>
      <a href="05-C-Advanced-bayesian-example.html">Advanced Bayesian model example</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W6
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-A-unconstrained-ordination.html">Unconstrained ordination</a>
    </li>
    <li>
      <a href="06-B-constrained-ordination.html">Constrained ordination</a>
    </li>
    <li>
      <a href="06-C-matrix-comparison.html">Comparing multivariate data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W8
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="08-A-mapping.html">Visualizing spatial data</a>
    </li>
    <li>
      <a href="08-B-spatial-regression.html">Spatial regression</a>
    </li>
    <li>
      <a href="08-C-spatial-ordination.html">Ordination approach to spatial analysis</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    W9
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="09-A-time-series.html">Time series</a>
    </li>
    <li>
      <a href="09-B-networks.html">Network analysis</a>
    </li>
    <li>
      <a href="09-C-occupancy-models.html">Occupancy models</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Introduction to ggplot2 and dplyr</h1>
<h4 class="author"><em>Jes</em></h4>

</div>


<p><strong>Assigned Reading:</strong></p>
<blockquote>
<p>Chapters 1.1, 1.2, 2, 3, 9 and 10 in: Wickham, H. 2009. <em>ggplot2: Elegant Graphics for Data Analysis.</em> Springer. DOI: <a href="https://link.springer.com/book/10.1007%2F978-0-387-98141-3">Stanford Full Text</a></p>
<p>The sections on mutating and filtering joins in <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/two-table.html">this vignette</a></p>
</blockquote>
<p><strong>Optional Reading:</strong></p>
<blockquote>
<p>Chapter 4 in: Wickham, H. 2009. <em>ggplot2: Elegant Graphics for Data Analysis.</em> Springer. DOI: <a href="https://link.springer.com/book/10.1007%2F978-0-387-98141-3">Stanford Full Text</a></p>
</blockquote>
<p>RStudio has produced several helpful <a href="https://www.rstudio.com/resources/cheatsheets/">cheatsheets</a> on data management and visualization with ggplot2 and tidyr. You may want to view or print one or several of them.</p>
<div id="before-class" class="section level3">
<h3>Before Class</h3>
<p>Download <a href="https://github.com/FukamiLab/BIO202/blob/master/code/02-A-tidyr-exercise-code.R">this R script</a> and save it in your code folder in the research project you created on the first day of class. You will need to click the ‘Raw’ button and then download the text file that appears. Be sure to remove the .txt extension that may have been added when you saved the file to your computer so that the file ends in .R. You may want to run the first part of the script that downloads the data to be sure that there will be no issues with data access during class.</p>
</div>
<div id="key-points" class="section level3">
<h3>Key Points</h3>
<ul>
<li>Data should be organized so that observations are in rows and attributes (variables) are in columns.</li>
<li>The <code>dplyr</code> package contains functions for manipulating data:
<ul>
<li><code>spread</code> and <code>gather</code> change the format of data tables.</li>
<li><code>separate</code> and <code>unite</code> change the format of variables.</li>
<li><code>filter</code> creates a subset of a data table.</li>
<li><code>mutate</code> creates new variables.</li>
<li><code>group_by</code> and <code>summarise</code> are used to summarize observations according to the values of their variables.</li>
</ul></li>
<li>The pipe function <code>%&gt;%</code> is used to link several operations together (e.g. function composition).</li>
<li>The <code>ggplot2</code> package is set of functions for visualizing data.
<ul>
<li>A ggplot2 plot contains three components: (1) the data, (2) the aesthetic mappings between variables and visual properties, and (3) layers describing how to display the observations.</li>
<li>The aesthetic mapping <code>aes</code> must minimally describe which variables define the plot coordinate space.</li>
<li><code>facet_wrap</code> creates plots for specified data subsets.</li>
</ul></li>
</ul>
</div>
<div id="manipulating-data-with-dplyr" class="section level3">
<h3>Manipulating data with dplyr</h3>
<p>Open the R project that you created on the first day of class. Then when R Studio opens, open the R script file that you downloaded for today’s class (see above). This script downloads a data set from our course website and then proceeds to manipulate and summarise the data. These data contain three tables from the Winter 2016 BIO46 class on lichen microorganisms:</p>
<ol style="list-style-type: decimal">
<li><strong>trees</strong> is a table where observations are trees on which lichens were collected. Columns describe each tree’s location, elevation, and species name.</li>
<li><strong>lichens</strong> is a table where each observation is a lichen that was collected. Columns describe the tree on which the lichen was collected, the team that collected the lichen, the collection date, and the lichen species name.</li>
<li><strong>algae</strong> is a table where each observation is a fragment of a lichen from which we attempted to amplify and sequence green algal ITS2 rDNA. We attempted to amplify algal DNA from four sections of each lichen in order to quantity algal diversity within a single lichen. If sequencing was successful then the DNA sequence is given along with a code identifying unique haplotype (GenotypeID). The TaxonID column indicates haplotypes that matched those reported in <a href="http://dw.doi.org/10.3732/ajb.0900276">Werth and Sork (2010)</a>.</li>
</ol>
<p>The R script explores algal diversity among lichens and trees. Working in pairs, execute each line of code and then add a comment (using <code>#</code>) above each line of code with a description of what the code does.</p>
<p>When you have finished commenting the code, try to complete the following challenge:</p>
<blockquote>
<p><strong>Challenge:</strong></p>
<p>How could you modify the code that creates the <code>lichenXgeno</code> table so that it displays the fraction of successful sequences belonging to each GenotypeID for each lichen.</p>
</blockquote>
</div>
<div id="plotting-data-with-ggplot2" class="section level3">
<h3>Plotting data with ggplot2</h3>
<p>Working with your partner, create the following plots using <code>ggplot2</code>:</p>
<ol style="list-style-type: decimal">
<li>A plot that maps the locations where each lichen was collected and colors the point by the Chao1 richness estimator.</li>
<li>Modify the above plot so that three plots are shown, one for each tree species.</li>
</ol>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
